### ISSUE TITLE
Grand Theft Auto: San Andreas, HW grass slightly off.

### ISSUE THREAD

**Issue by valcheeky** [PCSX2/pcsx2#1987]
PCSX2 version: latest git, v1.5.0-dev-2123-gf889720 2017-04-15 00:34:33
PCSX2 options: Default settings, MTVU enabled or disabled, issue is present with both settings.
Plugins used: GSdx AVX2, LilyPad-Scp-r5875, SPU2-X 2.0.0, cdvdGigaherz, Usbnul, Fwnull, DEv9null
OpenGL Hardware, default settings(native res, AF Off, etc), HW hack "Auto Flush" enabled.

Description of the issue:
Patches of grass are rendered slightly wrong in hardware mode, every hardware renderer is affected but OpenGL is recommended as it can run the game the best.
 Enabling settings "Allow 8-bit textures" "CRC hack levels" to "partial" or "full" "Accurate Date" "Blending Unit Accuracy" "Basic" or "High", "Minmapping" "Basic" or "Full", "Anisotropic Filtering 16x", make no difference whatsoever and issue remains.

How to reproduce the issue:
Simply enable HW Hack Auto Flush(#1612) and start a new game, head to any patch of grass and you compare switching between software and hardware mode.

Last known version to work: Probably never worked correctly?

PC specifications:
 CPU: Intel 4690K GPU: Gigabyte GTX 970 OS: Windows 10 64 Bit, latest Nvidia driver 384.76

How it looks in hardware mode
![hardware](https://user-images.githubusercontent.com/29753802/27720331-a6979b6c-5d1e-11e7-8488-7e44cdb8fd5f.png)
GS dump
https://www.dropbox.com/s/yg0itomveiufow1/hardware.xz?dl=0

How it look like in Software mode, please pay attention to the grass far away off the in distance.
![software](https://user-images.githubusercontent.com/29753802/27720407-f7b3c138-5d1e-11e7-85f9-d64c9d328057.png)
GS dump
https://www.dropbox.com/s/ac64l5odajqo82u/software.xz?dl=0

**Comment by refractionpcsx2** [comment]
Does it work without the auto flush hack?

**Comment by valcheeky** [comment]
Still happens with auto flush disabled, hard to see but the grass is the same  with or without auto flush enabled.
![gsdx_20170630124417](https://user-images.githubusercontent.com/29753802/27747716-fb02af02-5d91-11e7-9ecc-c6a4428bce2c.png)

**Comment by refractionpcsx2** [comment]
Ok I just wanted to check as I couldn't understand why turning it on is in your instructions. You should probably edit that :)

**Comment by valcheeky** [comment]
Without the hack this happens.
![gsdx_20170630132826](https://user-images.githubusercontent.com/29753802/27749200-2f983a24-5d98-11e7-93e9-02e1580d85a6.png)
Like a second frame layered on top offset by half the screen, it happens everywhere.

**Comment by gregory38** [comment]
Yes you need the autoflush hack for this game. Otherwise based on the description it could a trilinear filtering issue aka mipapping.

**Comment by gregory38** [comment]
Could you test SW renderer with SW mipmapping disabled.

**Comment by valcheeky** [comment]
sure, here it is with it mipmapping disabled
![softwaremipmappingdisabled](https://user-images.githubusercontent.com/29753802/27807748-82c8dc18-6008-11e7-89e6-ee2baaa3fdf3.png)

GS dump
https://www.dropbox.com/s/59zanxuos9se1an/softwaremipmappingdisabled.xz?dl=0

Mipmapping enabled
![softwaremipmappingenabled](https://user-images.githubusercontent.com/29753802/27807819-f289a276-6008-11e7-9c51-2eafaf4b0d2a.png)

GS dump
https://www.dropbox.com/s/brzisiaz0yg9qo1/softwaremipmappingenabled.xz?dl=0

**Comment by gregory38** [comment]
A single dump is enough, option doesn't impact the dump.

**Comment by gregory38** [comment]
And honestly I don't see the difference between all your screenshot. Could you post a bad one with a red circle where to look.

**Comment by valcheeky** [comment]
Theres no difference between both software renderer pics, so disabling/enabling mipmapping makes no difference for software renderer. It only happens in hardware mode..

**Comment by valcheeky** [comment]
Hi, I was messing around with this some more and noticed that hardware mode looks to be offset by 1 or 2 pixels to the left compared to software mode at native resolutions, it seems to affect textures too. At first I thought it was the "radiosity"(https://github.com/pcsx2/pcsx2/issues/600) postprocessing effect which was reported already but it happens to the whole image including the radiosity even without using the autoflush hack and **while** using native resolution, which looks like the original issue report fails to mention.

It's hard to appreciate in screenshots without doing extreme zoom which doing without losing quality is out of my scope but using fullscreen is very obvious when looking at advertising billboards and the right side edge of the screen looks transparent see-through

Another thing, using any custom resolution breaks everything but the hud and minimap(still looks really off  though), even the same resolution of 640x448(or 640x480 from the itnro) reported by software native makes it really bad.
![640 448](https://user-images.githubusercontent.com/29753802/28006639-d28d251c-6515-11e7-9e06-571ac7d7740b.png)

interestedly 640x449 looks better.
![640 449](https://user-images.githubusercontent.com/29753802/28006657-e43f7396-6515-11e7-9daa-26f9cb92ce34.png)

**Comment by gregory38** [comment]
Custom resolution doesn't support correct depth emulation.

Otherwise, your issue to be a rounding issue. The SW renderer use integer operation so most operations are truncated. GPU uses float operation so rounded.

**Comment by DIKH3D** [comment]
![image](https://user-images.githubusercontent.com/34287698/33634908-c2b8a568-da0d-11e7-9fe7-ac7aee9a7e44.png)
